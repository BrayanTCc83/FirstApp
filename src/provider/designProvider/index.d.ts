import { DesignProvider, useDesignContext } from ".";

/**
 * This method must be used inside a React.Component | React.Element, for its use is necessary
 * contain the element in a DesignProvider
 * 
 * @returns {Object} The design context for your app
 */
export function useDesignContext () : {
    theme : string,
    strTheme : string,
    mainColor : string,
    secondaryColor : string,
    fontSizes : number,
    fontColor : string,
    grayFontColor : string,
    whiteColor : string,
    width : string,
    height : string,
    changeTheme : (newTheme : string) => void,
    changeSchemaColor : (newSchema : string) => void
}

/**
 * The design context provider is necessary to access details commons about the style 
 * for the app, like theme (color schema), mode (white or dark background) and fonts styles 
 * (size, color, family) also width and height of the screen.
 * If you try to use that resources out the context provider, you can receive an error.
 *
 * @param {Object} props Default configuration for the app, theme and mode 
 * @returns {React.Component} React provider for app desing
 */
export function DesignProvider ( props : { theme ?: string , mode ?: string } ) : JSX.Element