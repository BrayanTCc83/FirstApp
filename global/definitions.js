//Files definition
const CONFIG = 'app-config.txt'
const AUTH = 'app-auth.txt'
const FILES_NAMES = {
    CONFIG,
    AUTH
}
//Styles definition
const GREEN_THEME = 'green';
const RED_THEME = 'red';
const DARK_MODE = 'dark';
const LIGHT_MODE = 'light';
const STYLE_DEFINITIONS = {
    GREEN_THEME,
    RED_THEME,
    DARK_MODE,
    LIGHT_MODE
};
//Languajes definition
const ES_LANG = "es";
const EN_LANG = "en";
const FR_LANG = "fr";
const LANG_DEFINITIONS = {
    ES_LANG,
    EN_LANG,
    FR_LANG
};
//Icons name definitions
const BELL_ICON = 'bell';
const GO_BACK_ICON = 'goback';
const CHAT_MESSAGE_ICON = 'chat_message';
const NEW_POST_ICON = 'feather';
const IMAGE_ICON = 'image';
const USER_ICON = 'user';
const PRIVACITY_ICON = 'lock';
const TOOLS_ICON = 'tools';
const CONFIGURATION_ICON = 'config';
const DROPDOWN_ICON = 'dropdown';
const SEND_MESSAGE_ICON = 'send';
const UNDEFINED_ICON = 'question';
const CAMERA = 'camera';
const SPIN = 'spin';
const FLASH = 'flash';
const RECORD = 'record';
const WHITE_BALANCE = 'white_balance';
const ICONS_DEFINITIONS = {
    BELL_ICON,
    GO_BACK_ICON,
    CHAT_MESSAGE_ICON,
    NEW_POST_ICON,
    IMAGE_ICON,
    USER_ICON,
    PRIVACITY_ICON,
    TOOLS_ICON,
    CONFIGURATION_ICON,
    DROPDOWN_ICON,
    SEND_MESSAGE_ICON,
    UNDEFINED_ICON,
    CAMERA,
    SPIN,
    FLASH,
    RECORD,
    WHITE_BALANCE
};
//Texts definition
const TEXT_SIZE_1=1;
const TEXT_SIZE_2=2;
const TEXT_SIZE_3=3;
const TEXT_SIZE_4=4;
const TEXT_SIZE_5=5;
const TEXT_DEFINITIONS = {
    TEXT_SIZE_1,
    TEXT_SIZE_2,
    TEXT_SIZE_3,
    TEXT_SIZE_4,
    TEXT_SIZE_5
}
//Screen defintion
const HOME_VIEW = "Home";
const CHAT_VIEW = "Chat";
const CHAT_LIST_VIEW = "ChatList";
const LANDING_VIEW = "Landing";
const LOAD_VIEW = "Load";
const LOGIN_VIEW = "Login";
const CONFIGURATION_VIEW = "Configuration";
const POST_VIEW = "Post";
const REGISTER_VIEW = "Register";
const USER_PROFILE_VIEW = "User";
const MULTIMEDIA = "Multimedia";
const POST_CREATE = 'Create post';
const CAMERA_TAKE = 'Camera';
const SCREEN_VIEWS = {
    HOME_VIEW,
    CHAT_LIST_VIEW,
    CHAT_VIEW,
    LANDING_VIEW,
    LOAD_VIEW,
    LOGIN_VIEW,
    CONFIGURATION_VIEW,
    POST_VIEW,
    REGISTER_VIEW,
    USER_PROFILE_VIEW,
    MULTIMEDIA,
    POST_CREATE,
    CAMERA_TAKE
}
export {
    TEXT_DEFINITIONS,
    STYLE_DEFINITIONS,
    LANG_DEFINITIONS,
    ICONS_DEFINITIONS,
    SCREEN_VIEWS,
    FILES_NAMES
};