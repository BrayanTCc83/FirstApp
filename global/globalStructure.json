{
    "vars" : {
        "colors" : {
            "darkTheme":"#434343",
            "lightTheme":"#FFFFFF",
            "greenStyle":{
                "main":"#1CCDAA",
                "secondary":"#d4ffee"
            },
            "redStyle":{
                "main":"#FF6060",
                "secondary":"#ffcccc"
            }
        },
        "fontSizes" : {
            "first":30,
            "second":25,
            "common":20,
            "info":17,
            "detail":12
        },
        "fontColors" : {
            "black":"#000000",
            "white":"#ffffff",
            "gray":"#aaaaaaaa"
        }
    }
}