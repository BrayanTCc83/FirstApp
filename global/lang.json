{
    "es":{
        "title":"Nombre de la app"
    }
}