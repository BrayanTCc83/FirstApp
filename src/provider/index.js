import React, { createContext, useContext, useMemo, useState } from 'react';
const AplicationContext = createContext();

export const AplicationProvider = (props) => {
    const [appTheme, setAppTheme] = useState({
        theme:'light', 
        style:'green'
    });

    const modifyTheme = (newTheme) => {
        setAppTheme(newTheme)
    };

    const value = useMemo(() => {
        return ({
            appTheme,
            modifyTheme
        })
    },[appTheme]);

    return <AplicationContext.Provider value={value} {...props} />
};

export const useAplicationContext = () => {
    const context = useContext(AplicationContext);
    if(!context){
        throw new Error("No es posible acceder al contexto de la aplicaci√≥n");
    }
};