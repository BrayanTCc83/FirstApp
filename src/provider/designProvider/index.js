import React, { createContext, useContext, useEffect, useMemo, useState } from 'react'
import { Dimensions } from 'react-native'

//Variables
import Variables from "../../../global/globalStructure.json"

//Definitions
import { STYLE_DEFINITIONS, FILES_NAMES } from "../../../global/definitions"
import useReactNativeFile from '../../hooks/useReactFileFile'

export const DesignContext = createContext()

export const DesignProvider = (props) => {
    const ReactFile = useReactNativeFile()

    const [theme, setTheme] = useState( Variables.vars.colors.lightTheme )
    const [strTheme, setStrTheme] = useState( STYLE_DEFINITIONS.LIGHT_MODE )
    const [mainColor, setMainColor] = useState( Variables.vars.colors.greenStyle.main )
    const [secondaryColor, setSecondaryColor] = useState( Variables.vars.colors.greenStyle.secondary )
    const [fontColor, setFontColor] = useState( Variables.vars.fontColors.black )
    const [strSchema, setStrSchema] = useState( STYLE_DEFINITIONS.GREEN_THEME )

    const [update, setUpdate] = useState(true)

    const getConfigTheme=()=>{
        ReactFile.fileExists( FILES_NAMES.CONFIG ).then(
            async (result) => {
                await setStrTheme( result['strTheme'] )
                await setFontColor( result['fontColor'] )
                await setTheme( result['theme'] )
                await setMainColor( result['mainColor'] )
                await setSecondaryColor( result['secondaryColor'] )
                await setStrSchema( result['strSchema'] )
                await setUpdate(false)
            }
        )
    }
    const changeTheme = async( newTheme ) => {
        var structure = {
            strTheme : newTheme,
            theme : newTheme === STYLE_DEFINITIONS.LIGHT_MODE ?
                Variables.vars.colors.lightTheme :
                Variables.vars.colors.darkTheme,
            fontColor: newTheme === STYLE_DEFINITIONS.LIGHT_MODE ?
                Variables.vars.fontColors.black :
                Variables.vars.fontColors.white
        }
        await ReactFile.fileExists( FILES_NAMES.CONFIG ).then(
            () => ReactFile.editFile( FILES_NAMES.CONFIG, structure )
        ).catch(
            () => ReactFile.writeFile( FILES_NAMES.CONFIG, structure )
        )
        await setUpdate(true)
    }
    const changeSchemaColor = async( newSchema ) => {
        var structure = {
            strSchema : newSchema,
            mainColor : newSchema === STYLE_DEFINITIONS.GREEN_THEME ?
                Variables.vars.colors.greenStyle.main :
                Variables.vars.colors.redStyle.main,
            secondaryColor: newSchema === STYLE_DEFINITIONS.GREEN_THEME ?
                Variables.vars.colors.greenStyle.secondary :
                Variables.vars.colors.redStyle.secondary
        }
        await ReactFile.fileExists( FILES_NAMES.CONFIG ).then(
            () => ReactFile.editFile( FILES_NAMES.CONFIG, structure )
        ).catch(
            () => ReactFile.writeFile( FILES_NAMES.CONFIG, structure )
        )
        await setUpdate(true)
    }   

    useEffect(()=>{
        if( update ){
            return getConfigTheme()
        }
    })

    const grayFontColor = Variables.vars.fontColors.gray

    const whiteColor = Variables.vars.fontColors.white

    const width = Dimensions.get('window').width
    
    const height = Dimensions.get('window').height

    const fontSizes = Variables.vars.fontSizes

    const value = useMemo(() => {
        return ({
            theme,
            mainColor,
            secondaryColor,
            strTheme,
            changeTheme,
            changeSchemaColor,
            width,
            height,
            fontSizes,
            fontColor,
            grayFontColor,
            whiteColor,
            strSchema
        })
    }, [ theme, width, height, secondaryColor ] );
    return <DesignContext.Provider value={value} {...props} />
};

export const useDesignContext = () => {
    const context = useContext(DesignContext);
    if(!context){
        throw new Error("No es posible acceder al contexto de la aplicaci√≥n");
    }
    return context
};