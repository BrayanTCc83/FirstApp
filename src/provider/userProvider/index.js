import React, { useContext, useEffect, useMemo, useState } from 'react'
import Database from "../../../database"
import { SecurityHandler } from '../../../functions'
import useReactFile from "../../hooks/useReactFileFile"
import { FILES_NAMES } from "../../../global/definitions"
import { useHandlerData } from '../../hooks/handlers'
const UserContext = React.createContext()
var save = false
const UserProvider = ( props ) => {
    const [ isFileAuthRead, setFileRead ] = useState( save )
    const [ isUser, setUser ] = useState( false )
    const [ userData, setUserData, cleanData ] = useHandlerData()
    const UserDocument = Database( 'users' )
    const keys = UserDocument.getKeys()
    const ReactFile = useReactFile()

    const readUserFile = () => {
        ReactFile.fileExists( FILES_NAMES.AUTH ).catch((err)=>console.log(err))
            .then((value)=>{
                if( typeof value === 'object' ){
                    const token = value.token ? value.token : ""
                    if ( token !== "" ){
                        if( SecurityHandler.validateToken( token ) ){
                            const accessKey = SecurityHandler.getTokenData( token )[keys[0]]
                            setUserData( accessKey, "key" )
                            UserDocument.getOneValue( accessKey ,keys[8], (hash)=>{
                                setUser( SecurityHandler.getAccess( token, hash ) )
                                UserDocument.getValues( accessKey, ( key, value, index )=>{
                                    if( ! ( key === keys[8] || key === keys[9] ) ){
                                        setUserData( value, key )
                                    }
                                } )
                            })
                        }
                    }
                }
            })
            .then(()=>{
                save = true
                setFileRead(true)
            })
            .catch((err)=>{
                save = false
                setFileRead(true)
                console.log(err)
            })
    }

    const login = ( data ) => { 
        let forToken = {
            [keys[0]] : data[keys[0]],
            [keys[7]] : data[keys[7]]
        }
        const token = SecurityHandler.generateToken( forToken, data[keys[8]] )
        console.log( accessKey )
        const UserFile = {
            'token' : token
        }
        ReactFile.writeFile( FILES_NAMES.AUTH, UserFile )
    }
    const logout = () => {
        setUser(false)
        cleanData()
        ReactFile.writeFile( FILES_NAMES.AUTH, { token : '' } )
    }
    const register = ( newUser ) => {
        if( UserDocument.pushElement( newUser ) ) {
            login( newUser )
        }
    }

    const value = useMemo( () => {
        return ({
            isUser,
            userData,
            isFileAuthRead,
            register,
            login,
            logout,
            readUserFile
        })
    }, [ isUser, userData, isFileAuthRead ] )
    return (
        <UserContext.Provider value = {value} {...props} >
            {props.children}
        </UserContext.Provider>
    )
}
const useUserContext = () => {
    const context = useContext( UserContext )
    if( ! context ){
        throw new Error("The user context must be used on a provider, this context provide user information like token and user data")
    }
    return context
}
export {
    UserProvider,
    UserContext,
    useUserContext
}