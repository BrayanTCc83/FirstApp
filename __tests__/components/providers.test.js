import 'react-native';
import React from 'react';

import renderer from 'react-test-renderer';

import { DesignProvider } from "../../src/provider/designProvider"
import { STYLE_DEFINITIONS } from "../../global/definitions"

import Button from "../../src/components/button";
import TextView from "../../src/components/textView";
import Input from "../../src/components/input";
import Message from "../../src/components/message";
import SwitchButton from "../../src/components/switchButton";
import Icon from "../../src/components/icon";
import ChatSelector from "../../src/components/chatSelector";
import Dropdown from "../../src/components/dropdown";
import ViewContainer from "../../src/components/viewContainer";

const DEFAULT_PROVIDER_PROPS = {
    theme : STYLE_DEFINITIONS.DARK_MODE,
    mode : STYLE_DEFINITIONS.GREEN_THEME
}

const designRender = ( ui, { providerProps } ) => {
    return renderer.create(
        <DesignProvider {...providerProps} >
            {ui}
        </DesignProvider>
    )
}

it('Button tested with design provider', () => {
    const BUTTON_TEST = designRender( <Button/> , DEFAULT_PROVIDER_PROPS ).toJSON();
    expect( BUTTON_TEST ).toMatchSnapshot();
});

it('Text tested with design provider', () => {
    const TEXT_VIEW_TEST = designRender( <TextView/>, DEFAULT_PROVIDER_PROPS ).toJSON();
    expect( TEXT_VIEW_TEST ).toMatchSnapshot();
});

it('View container tested with design provider', () => {
    const VIEW_CONTAINER_TEST =  designRender( <ViewContainer/>, DEFAULT_PROVIDER_PROPS ).toJSON();
    expect( VIEW_CONTAINER_TEST ).toMatchSnapshot();
});

it('Input tested with desing provider', () => {
    const INPUT_TEST = designRender( <Input/>, DEFAULT_PROVIDER_PROPS ).toJSON();
    expect( INPUT_TEST ).toMatchSnapshot();
});

it('Message component tested with desing provider', () => {
    const MESSAGE_COMPONENT_TEST = designRender( <Message/>, DEFAULT_PROVIDER_PROPS ).toJSON();
    expect( MESSAGE_COMPONENT_TEST ).toMatchSnapshot();
});

it('Switch button tested with desing provider', () => {
    const SWITCH_BUTTON_TEST = designRender( <SwitchButton/>, DEFAULT_PROVIDER_PROPS ).toJSON();
    expect( SWITCH_BUTTON_TEST ).toMatchSnapshot();
});

it('Icon component tested with desing provider', () => {
    const ICON_COMPONENT_TEST = designRender( <Icon/>, DEFAULT_PROVIDER_PROPS ).toJSON();
    expect( ICON_COMPONENT_TEST ).toMatchSnapshot();
});

it('Chat selector tested with desing provider', () => {
    const CHAR_SELECTOR_TEST = designRender( <ChatSelector/>, DEFAULT_PROVIDER_PROPS ).toJSON();
    expect( CHAR_SELECTOR_TEST ).toMatchSnapshot();
});
