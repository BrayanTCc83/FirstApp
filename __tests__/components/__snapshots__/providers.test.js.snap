// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button tested with design provider 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#1CCDAA",
      "borderColor": "#1CCDAA",
      "borderRadius": 150,
      "borderWidth": 1,
      "left": 75,
      "marginVertical": 10,
      "maxWidth": 600,
      "padding": 12,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#ffffff",
        "fontSize": 23,
        "fontWeight": "bold",
        "marginHorizontal": 0,
        "textAlign": "center",
      }
    }
  >
    Button
  </Text>
</View>
`;

exports[`Icon component tested with desing provider 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#1CCDAA",
      "borderRadius": 187.5,
      "height": 375,
      "left": 187.5,
      "top": 222.33333333333334,
      "width": 375,
    }
  }
>
  <View
    style={
      Object {
        "color": "#1CCDAA",
        "height": 50,
        "left": 8,
        "top": 8,
        "width": 50,
      }
    }
  />
</View>
`;

exports[`Input tested with desing provider 1`] = `
<TextInput
  allowFontScaling={true}
  onChange={[Function]}
  placeholder="Input"
  rejectResponderTermination={true}
  style={
    Object {
      "backgroundColor": "transparent",
      "borderColor": "#1CCDAA",
      "borderRadius": 15,
      "borderWidth": 1,
      "fontSize": 20,
      "fontWeight": "bold",
      "left": 75,
      "marginVertical": 10,
      "maxWidth": 600,
      "paddingHorizontal": 15,
      "zIndex": 2,
    }
  }
  underlineColorAndroid="transparent"
/>
`;

exports[`Message component tested with desing provider 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#CDFFB6",
      "borderBottomLeftRadius": 0,
      "borderColor": "#1CCDAA",
      "borderRadius": 15,
      "borderWidth": 2,
      "marginHorizontal": 15,
      "marginVertical": 7,
      "padding": 6,
      "paddingBottom": 15,
      "width": 500,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 20,
        "fontWeight": "normal",
        "marginHorizontal": 0,
        "textAlign": "justify",
      }
    }
  >
    Message
  </Text>
</View>
`;

exports[`Switch button tested with desing provider 1`] = `
<View
  style={
    Object {
      "height": 50,
      "marginHorizontal": 75,
      "marginVertical": 10,
      "width": 600,
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#CDFFB6",
        "borderColor": "#1CCDAA",
        "borderRadius": 50,
        "borderWidth": 1,
        "height": 50,
        "width": 100,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#1CCDAA",
          "borderRadius": 20,
          "height": 40,
          "left": 5,
          "top": 4,
          "width": 40,
        }
      }
    />
  </View>
  <Text
    style={
      Object {
        "color": "#000000",
        "fontSize": 20,
        "fontWeight": "bold",
        "left": 105,
        "marginHorizontal": 0,
        "textAlign": "left",
        "top": -40,
        "width": 375,
      }
    }
  >
    Default value 1
  </Text>
</View>
`;

exports[`Text tested with design provider 1`] = `
<Text
  style={
    Object {
      "color": "#000000",
      "fontSize": 30,
      "fontWeight": "bold",
      "marginHorizontal": 0,
      "marginVertical": 50,
      "textAlign": "center",
    }
  }
>
  Text
</Text>
`;

exports[`View container tested with design provider 1`] = `
<RCTScrollView
  scrollEnabled={false}
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "height": 1334,
      "width": 750,
    }
  }
>
  <View />
</RCTScrollView>
`;
